list/schema/schema.go

```diff
@@ -3,5 +3,195 @@
+func (s Schema) AttributeAtPath(ctx context.Context, p path.Path) (fwschema.Attribute, diag.Diagnostics) {
+	return fwschema.SchemaAttributeAtPath(ctx, s, p)
+}
+
+// AttributeAtPath returns the Attribute at the passed path. If the path points
+// to an element or attribute of a complex type, rather than to an Attribute,
+// it will return an ErrPathInsideAtomicAttribute error.
+func (s Schema) AttributeAtTerraformPath(ctx context.Context, p *tftypes.AttributePath) (fwschema.Attribute, error) {
+	return fwschema.SchemaAttributeAtTerraformPath(ctx, s, p)
+}
+
+// GetAttributes returns the Attributes field value.
+func (s Schema) GetAttributes() map[string]fwschema.Attribute {
+	return schemaAttributes(s.Attributes)
+}
+
+// GetBlocks returns the Blocks field value.
+func (s Schema) GetBlocks() map[string]fwschema.Block {
+	return schemaBlocks(s.Blocks)
+}
+
+// GetDeprecationMessage returns the DeprecationMessage field value.
+func (s Schema) GetDeprecationMessage() string {
+	return s.DeprecationMessage
+}
+
+// GetDescription returns the Description field value.
+func (s Schema) GetDescription() string {
+	return s.Description
+}
+
+// GetMarkdownDescription returns the MarkdownDescription field value.
+func (s Schema) GetMarkdownDescription() string {
+	return s.MarkdownDescription
+}
+
+// GetVersion returns the Version field value.
+func (s Schema) GetVersion() int64 {
+	return s.Version
+}
+
+// Type returns the framework type of the schema.
+func (s Schema) Type() attr.Type {
+	return fwschema.SchemaType(s)
+}
+
+// TypeAtPath returns the framework type at the given schema path.
+func (s Schema) TypeAtPath(ctx context.Context, p path.Path) (attr.Type, diag.Diagnostics) {
+	return fwschema.SchemaTypeAtPath(ctx, s, p)
+}
+
+// TypeAtTerraformPath returns the framework type at the given tftypes path.
+func (s Schema) TypeAtTerraformPath(ctx context.Context, p *tftypes.AttributePath) (attr.Type, error) {
+	return fwschema.SchemaTypeAtTerraformPath(ctx, s, p)
+}
+
+// Validate verifies that the schema is not using a reserved field name for a top-level attribute.
+//
+// Deprecated: Use the ValidateImplementation method instead.
+func (s Schema) Validate() diag.Diagnostics {
+	return s.ValidateImplementation(context.Background())
+}
+
+// ValidateImplementation contains logic for validating the provider-defined
+// implementation of the schema and underlying attributes and blocks to prevent
+// unexpected errors or panics. This logic runs during the
+// ValidateResourceConfig RPC, or via provider-defined unit testing, and should
+// never include false positives.
+func (s Schema) ValidateImplementation(ctx context.Context) diag.Diagnostics {
+	var diags diag.Diagnostics
+
+	for attributeName, attribute := range s.GetAttributes() {
+		req := fwschema.ValidateImplementationRequest{
+			Name: attributeName,
+			Path: path.Root(attributeName),
+		}
+
+		diags.Append(fwschema.IsReservedResourceAttributeName(req.Name, req.Path)...)
```

Comment:
Just a reminder (since it's also relevant for state storage), we should double check if there are any additional fields that should be protect from shadowing that aren't the same as managed resources.

For example, `state_store` blocks will only need to ensure there is no `provider` attribute in the schema.